File: .agents/skills/desktop-manage-git-lifecycle/scripts/git_lifecycle_test_support.py
"""为 Git 生命周期黑盒回归提供共享的隔离仓库夹具。"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
import importlib.util
import json
import os
from pathlib import Path
import subprocess
import sys
import tempfile
import unittest


SCRIPT = Path(__file__).with_name("git_lifecycle.py")
RELEASE_CONTEXT_HELPER = (
    SCRIPT.parents[2] / "desktop-prepare-release" / "scripts" / "release_context.py"
)
SHANGHAI = timezone(timedelta(hours=8), name="Asia/Shanghai")


def load_lifecycle_module() -> object:
    """加载被测 helper 模块，供无法由真实 Git 稳定制造的异常路径做受控注入。"""

    module_name = "_agent_first_git_lifecycle_test_target"
    specification = importlib.util.spec_from_file_location(module_name, SCRIPT)
    if specification is None or specification.loader is None:
        raise RuntimeError("Git lifecycle test module cannot be loaded.")
    module = importlib.util.module_from_spec(specification)
    sys.modules[module_name] = module
    script_directory = str(SCRIPT.parent)
    sys.path.insert(0, script_directory)
    try:
        specification.loader.exec_module(module)
    finally:
        sys.path.remove(script_directory)
    return module


LIFECYCLE = load_lifecycle_module()


class GitLifecycleTestCase(unittest.TestCase):
    """建立真实临时仓库并封装生命周期测试共用操作。"""

    def setUp(self) -> None:
        """为每个场景建立自动回收的隔离文件系统根。"""

        self.temporary = tempfile.TemporaryDirectory()
        self.root = Path(self.temporary.name)

    def tearDown(self) -> None:
        """回收测试仓库、远端和 Worktree，不触碰真实用户数据。"""

        self.temporary.cleanup()

    def git(
        self,
        cwd: Path,
        *arguments: str,
        check: bool = True,
    ) -> subprocess.CompletedProcess[str]:
        """以测试专用非交互环境执行 Git，并在意外失败时展示夹具诊断。"""

        environment = os.environ.copy()
        environment["GIT_TERMINAL_PROMPT"] = "0"
        environment["GCM_INTERACTIVE"] = "Never"
        result = subprocess.run(
            ["git", "-C", str(cwd), *arguments],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            encoding="utf-8",
            errors="replace",
            env=environment,
            check=False,
        )
        if check and result.returncode != 0:
            self.fail(f"git {' '.join(arguments)} failed: {result.stderr}")
        return result

    def initialize_repository(self, *, remote: bool) -> tuple[Path, Path | None]:
        """创建带首个 main 提交的仓库，并按场景选择本地 bare 远端。"""

        repository = self.root / "repository"
        repository.mkdir()
        self.git(repository, "init", "-b", "main")
        self.git(repository, "config", "user.name", "Lifecycle Test")
        self.git(repository, "config", "user.email", "lifecycle@example.invalid")
        helper = repository / ".agents/skills/desktop-prepare-release/scripts/release_context.py"
        helper.parent.mkdir(parents=True)
        helper.write_bytes(RELEASE_CONTEXT_HELPER.read_bytes())
        (repository / "base.txt").write_text("base\n", encoding="utf-8")
        self.git(repository, "add", "base.txt", str(helper.relative_to(repository)))
        self.git(repository, "commit", "-m", "initial")
        if not remote:
            return repository, None
        bare = self.root / "remote.git"
        bare.mkdir()
        self.git(bare, "init", "--bare")
        self.git(bare, "symbolic-ref", "HEAD", "refs/heads/main")
        self.git(repository, "remote", "add", "origin", str(bare))
        self.git(repository, "push", "-u", "origin", "main")
        return repository, bare

    def add_bare_remote(self, repository: Path, name: str, default_branch: str) -> Path:
        """增加具有独立默认分支的真实 bare 远端，并以当前 main 初始化它。"""

        bare = self.root / f"{name}.git"
        bare.mkdir()
        self.git(bare, "init", "--bare")
        self.git(bare, "symbolic-ref", "HEAD", f"refs/heads/{default_branch}")
        self.git(repository, "remote", "add", name, str(bare))
        self.git(
            repository,
            "push",
            name,
            f"refs/heads/main:refs/heads/{default_branch}",
        )
        return bare

    def helper(
        self,
        repository: Path,
        *arguments: str,
        success: bool = True,
        process_cwd: Path | None = None,
    ) -> tuple[dict[str, object], subprocess.CompletedProcess[str]]:
        """运行 helper，验证标准输出始终是唯一一行 JSON 和预期退出状态。"""

        result = subprocess.run(
            [sys.executable, str(SCRIPT), *arguments, "--project-root", str(repository)],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            encoding="utf-8",
            errors="replace",
            cwd=process_cwd,
            check=False,
        )
        self.assertEqual(result.stderr, "")
        self.assertEqual(len(result.stdout.splitlines()), 1, result.stdout)
        payload = json.loads(result.stdout)
        if success:
            self.assertEqual(result.returncode, 0, payload)
            self.assertNotEqual(payload.get("status"), "error")
        else:
            self.assertNotEqual(result.returncode, 0, payload)
            self.assertEqual(payload.get("status"), "error")
        return payload, result

    def commit_file(self, repository: Path, name: str, content: str) -> str:
        """在指定 Worktree 提交一个可观察文件并返回新提交 OID。"""

        (repository / name).write_text(content, encoding="utf-8")
        self.git(repository, "add", name)
        self.git(repository, "commit", "-m", f"add {name}")
        return self.git(repository, "rev-parse", "HEAD").stdout.strip()

    def state(self, repository: Path) -> dict[str, object]:
        """从 Git common-dir 读取 helper 的未跟踪生命周期状态。"""

        common = self.git(
            repository,
            "rev-parse",
            "--path-format=absolute",
            "--git-common-dir",
        ).stdout.strip()
        return json.loads(
            (Path(common) / "agent-first-harness" / "git-lifecycle.json").read_text()
        )

    def prepare_release_context(
        self,
        repository: Path,
        *,
        version: str,
        date: str,
        git_publication: str,
        remote: str | None,
        default_branch: str = "main",
        summary: str = "发布上下文测试",
    ) -> tuple[str, str]:
        """生成权威 helper 的规范上下文，提交后返回摘要和元数据 HEAD。"""

        source_head = self.git(repository, "rev-parse", "HEAD").stdout.strip()
        command = [
            sys.executable,
            str(repository / ".agents/skills/desktop-prepare-release/scripts/release_context.py"),
            "write",
            "--project-root",
            str(repository),
            "--source-head",
            source_head,
            "--version",
            version,
            "--release-date",
            datetime.strptime(date, "%Y%m%d").strftime("%Y-%m-%d"),
            "--review-selection",
            "disabled",
            "--scope-base",
            source_head,
            "--scope-diff-sha256",
            "0" * 64,
            "--review-reason",
            summary,
            "--review-remaining-risk",
            "测试不执行正式发布语义审查",
            "--macos-signing-selection",
            "not-applicable",
            "--macos-signing-source",
            "not-applicable",
        ]
        if git_publication == "local":
            command.extend(("--local-only", "--default-branch", default_branch))
        else:
            assert remote is not None
            command.extend(("--remote", remote))
        written = subprocess.run(
            command,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            encoding="utf-8",
            errors="replace",
            check=False,
        )
        self.assertEqual(written.returncode, 0, written.stderr)
        payload = json.loads(written.stdout)
        self.git(repository, "add", ".harness/release-context.json")
        self.git(repository, "commit", "-m", "chore(release): bind release context")
        return (
            str(payload["releaseContextSha256"]),
            self.git(repository, "rev-parse", "HEAD").stdout.strip(),
        )

    def local_branch_exists(self, repository: Path, branch: str) -> bool:
        """精确判断本地分支是否存在。"""

        return (
            self.git(
                repository,
                "show-ref",
                "--verify",
                "--quiet",
                f"refs/heads/{branch}",
                check=False,
            ).returncode
            == 0
        )

    def remote_branch_exists(self, repository: Path, branch: str) -> bool:
        """精确判断 origin 上的分支是否存在。"""

        return bool(
            self.git(
                repository,
                "ls-remote",
                "--heads",
                "origin",
                f"refs/heads/{branch}",
            ).stdout.strip()
        )

    def install_hook(self, bare: Path, body: str) -> None:
        """安装测试专用 pre-receive hook，以观察标签和清理的远端顺序。"""

        hook = bare / "hooks" / "pre-receive"
        hook.write_text("#!/bin/sh\nset -eu\n" + body, encoding="utf-8")
        hook.chmod(0o755)

File: .agents/skills/desktop-manage-git-lifecycle/scripts/git_lifecycle_test_support.mjs
/** Git 生命周期黑盒回归的隔离仓库夹具。 */

import assert from "node:assert/strict";
import {
  chmodSync,
  copyFileSync,
  mkdirSync,
  mkdtempSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { spawn, spawnSync } from "node:child_process";

const SCRIPT_DIRECTORY = dirname(fileURLToPath(import.meta.url));
export const SCRIPT = join(SCRIPT_DIRECTORY, "git_lifecycle.mjs");
export const RELEASE_CONTEXT_HELPER = resolve(
  SCRIPT_DIRECTORY,
  "../../desktop-prepare-release/scripts/release_context.mjs",
);

/** 以测试专用非交互环境执行命令。 */
export function run(command, args, options = {}) {
  const result = spawnSync(command, args, {
    cwd: options.cwd,
    env: options.env ?? process.env,
    encoding: "utf8",
    input: "",
    maxBuffer: 32 * 1024 * 1024,
  });
  if (result.error) throw result.error;
  if (options.check !== false && result.status !== 0) {
    assert.fail(`${command} ${args.join(" ")} failed: ${result.stderr}`);
  }
  return result;
}

/** 建立可自动回收的测试根及共享操作。 */
export function lifecycleFixture() {
  const temporary = mkdtempSync(join(tmpdir(), "git-lifecycle-"));
  const env = {
    ...process.env,
    GIT_CONFIG_GLOBAL: join(temporary, "global.gitconfig"),
    GIT_CONFIG_NOSYSTEM: "1",
    GIT_TERMINAL_PROMPT: "0",
    GCM_INTERACTIVE: "Never",
    SSH_ASKPASS_REQUIRE: "never",
  };
  const git = (cwd, ...args) => run("git", ["-C", cwd, ...args], { env });
  const gitUnchecked = (cwd, ...args) => run("git", ["-C", cwd, ...args], { env, check: false });
  const helper = (repository, args, { success = true, cwd } = {}) => {
    const result = run(process.execPath, [SCRIPT, ...args, "--project-root", repository], {
      env,
      cwd,
      check: false,
    });
    assert.equal(result.stderr, "");
    assert.equal(result.stdout.trimEnd().split("\n").length, 1, result.stdout);
    const payload = JSON.parse(result.stdout);
    if (success) {
      assert.equal(result.status, 0, JSON.stringify(payload));
      assert.notEqual(payload.status, "error");
    } else {
      assert.notEqual(result.status, 0, JSON.stringify(payload));
      assert.equal(payload.status, "error");
    }
    return { payload, result };
  };
  const initializeRepository = ({ remote }) => {
    const repository = join(temporary, "repository");
    mkdirSync(repository);
    git(repository, "init", "--quiet", "-b", "main");
    git(repository, "config", "user.name", "Lifecycle Test");
    git(repository, "config", "user.email", "lifecycle@example.invalid");
    const helperPath = join(repository, ".agents/skills/desktop-prepare-release/scripts/release_context.mjs");
    mkdirSync(dirname(helperPath), { recursive: true });
    copyFileSync(RELEASE_CONTEXT_HELPER, helperPath);
    writeFileSync(join(repository, "base.txt"), "base\n", "utf8");
    git(repository, "add", "base.txt", ".agents/skills/desktop-prepare-release/scripts/release_context.mjs");
    git(repository, "commit", "--quiet", "-m", "initial");
    if (!remote) return { repository, bare: null };
    const bare = join(temporary, "remote.git");
    mkdirSync(bare);
    git(bare, "init", "--quiet", "--bare");
    git(bare, "symbolic-ref", "HEAD", "refs/heads/main");
    git(repository, "remote", "add", "origin", bare);
    git(repository, "push", "--quiet", "-u", "origin", "main");
    return { repository, bare };
  };
  const addBareRemote = (repository, name, defaultBranch) => {
    const bare = join(temporary, `${name}.git`);
    mkdirSync(bare);
    git(bare, "init", "--quiet", "--bare");
    git(bare, "symbolic-ref", "HEAD", `refs/heads/${defaultBranch}`);
    git(repository, "remote", "add", name, bare);
    git(repository, "push", "--quiet", name, `refs/heads/main:refs/heads/${defaultBranch}`);
    return bare;
  };
  const commitFile = (repository, name, content) => {
    writeFileSync(join(repository, name), content, "utf8");
    git(repository, "add", name);
    git(repository, "commit", "--quiet", "-m", `add ${name}`);
    return git(repository, "rev-parse", "HEAD").stdout.trim();
  };
  const state = (repository) => {
    const common = git(repository, "rev-parse", "--path-format=absolute", "--git-common-dir").stdout.trim();
    return JSON.parse(readFileSync(join(common, "agent-first-harness/git-lifecycle.json"), "utf8"));
  };
  const prepareReleaseContext = (
    repository,
    { version, date, gitPublication, remote, defaultBranch = "main", summary = "发布上下文测试" },
  ) => {
    const sourceHead = git(repository, "rev-parse", "HEAD").stdout.trim();
    const command = [
      join(repository, ".agents/skills/desktop-prepare-release/scripts/release_context.mjs"),
      "write", "--project-root", repository,
      "--source-head", sourceHead,
      "--version", version,
      "--release-date", `${date.slice(0, 4)}-${date.slice(4, 6)}-${date.slice(6)}`,
      "--review-selection", "disabled",
      "--scope-base", sourceHead,
      "--scope-diff-sha256", "0".repeat(64),
      "--review-reason", summary,
      "--review-remaining-risk", "测试不执行正式发布语义审查",
      "--macos-signing-selection", "not-applicable",
      "--macos-signing-source", "not-applicable",
    ];
    if (gitPublication === "local") command.push("--local-only", "--default-branch", defaultBranch);
    else command.push("--remote", remote);
    const written = run(process.execPath, command, { env, check: false });
    assert.equal(written.status, 0, written.stderr);
    const payload = JSON.parse(written.stdout);
    git(repository, "add", ".harness/release-context.json");
    git(repository, "commit", "--quiet", "-m", "chore(release): bind release context");
    return {
      digest: payload.releaseContextSha256,
      head: git(repository, "rev-parse", "HEAD").stdout.trim(),
    };
  };
  const localBranchExists = (repository, branch) =>
    gitUnchecked(repository, "show-ref", "--verify", "--quiet", `refs/heads/${branch}`).status === 0;
  const remoteBranchExists = (repository, remote, branch) =>
    git(repository, "ls-remote", "--heads", remote, `refs/heads/${branch}`).stdout.trim() !== "";
  const installHook = (bare, body) => {
    const hook = join(bare, "hooks/pre-receive");
    writeFileSync(hook, `#!/bin/sh\nset -eu\n${body}`, "utf8");
    chmodSync(hook, 0o755);
  };
  const cleanup = () => rmSync(temporary, { recursive: true, force: true });
  return {
    temporary, env, git, gitUnchecked, helper, initializeRepository, addBareRemote,
    commitFile, state, prepareReleaseContext, localBranchExists, remoteBranchExists,
    installHook, cleanup,
  };
}

/** 异步启动一个 lifecycle CLI，供锁并发测试。 */
export function spawnHelper(repository, args, env) {
  return spawn(process.execPath, [SCRIPT, ...args, "--project-root", repository], {
    env,
    stdio: ["ignore", "pipe", "pipe"],
  });
}

/** 收集异步子进程并返回输出。 */
export function collectProcess(child) {
  return new Promise((accept, reject) => {
    let stdout = "";
    let stderr = "";
    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (chunk) => { stdout += chunk; });
    child.stderr.on("data", (chunk) => { stderr += chunk; });
    child.once("error", reject);
    child.once("close", (status) => accept({ status, stdout, stderr }));
  });
}
